File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/notes.rb
# Campaign notes: player-authored, role-filtered by visibility. The DM sees
# every note; players see all "party" notes plus only their own "private"
# ones. Ownership is derived from the authenticated actor, never the body.

def play_note_payload(row)
  {
    note_id: row['note_id'],
    text: row['text'],
    visibility: row['visibility'],
    owner: row['owner']
  }
end

def valid_note_visibility?(visibility)
  %w[private party].include?(visibility)
end

def next_play_note_sequence(campaign_id)
  db.execute(
    'SELECT COALESCE(MAX(sequence), 0) + 1 AS n FROM play_notes WHERE campaign_id = ?',
    [campaign_id]
  ).first['n']
end

def find_play_note!(campaign_id, note_id)
  note = db.execute(
    'SELECT * FROM play_notes WHERE campaign_id = ? AND note_id = ?',
    [campaign_id, note_id]
  ).first
  halt 404, { error: 'note not found' }.to_json unless note
  note
end

post '/v1/play/campaigns/:id/notes' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  note_id = body['note_id']
  text = body['text']
  visibility = body['visibility']

  halt 400, { error: 'invalid note_id' }.to_json unless note_id.is_a?(String) && !note_id.empty?
  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?
  halt 400, { error: 'invalid visibility' }.to_json unless valid_note_visibility?(visibility)

  existing = db.execute(
    'SELECT 1 FROM play_notes WHERE campaign_id = ? AND note_id = ?',
    [campaign['id'], note_id]
  ).first
  halt 409, { error: 'note_id already exists' }.to_json if existing

  sequence = next_play_note_sequence(campaign['id'])
  db.execute(
    'INSERT INTO play_notes (campaign_id, sequence, note_id, text, visibility, owner) VALUES (?, ?, ?, ?, ?, ?)',
    [campaign['id'], sequence, note_id, text, visibility, user['username']]
  )

  status 201
  { note_id: note_id, text: text, visibility: visibility, owner: user['username'] }.to_json
end

get '/v1/play/campaigns/:id/notes' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  rows = db.execute(
    'SELECT * FROM play_notes WHERE campaign_id = ? ORDER BY sequence ASC',
    [campaign['id']]
  )

  records = rows.map { |row| play_note_payload(row) }
  unless is_owner
    records = records.select { |n| n[:visibility] == 'party' || n[:owner] == user['username'] }
  end

  { notes: records }.to_json
end

get '/v1/play/campaigns/:id/notes/:note_id' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  is_owner = require_play_participant!(campaign, user, 'not a campaign member')

  note = find_play_note!(campaign['id'], params[:note_id])
  halt 403, { error: 'forbidden' }.to_json if !is_owner && note['visibility'] == 'private' && note['owner'] != user['username']

  play_note_payload(note).to_json
end

put '/v1/play/campaigns/:id/notes/:note_id' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  note = find_play_note!(campaign['id'], params[:note_id])
  halt 403, { error: 'not the note owner' }.to_json unless note['owner'] == user['username']

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  text = body['text']
  visibility = body['visibility']

  halt 400, { error: 'invalid text' }.to_json unless text.is_a?(String) && !text.empty?
  halt 400, { error: 'invalid visibility' }.to_json unless valid_note_visibility?(visibility)

  db.execute(
    'UPDATE play_notes SET text = ?, visibility = ? WHERE campaign_id = ? AND note_id = ?',
    [text, visibility, campaign['id'], params[:note_id]]
  )

  updated = find_play_note!(campaign['id'], params[:note_id])
  play_note_payload(updated).to_json
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/notes.ts
/**
 * Campaign notes with role-filtered privacy. Any campaign member may create
 * a note owned by themselves; the DM can read every note, but players only
 * see `party` notes plus their own `private` notes.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { authenticate, isActor, isApiResult, findCampaign, requireParticipant } from './shared.ts';

interface NoteRow {
  campaign_id: string;
  sequence: number;
  note_id: string;
  text: string;
  visibility: string;
  owner: string;
}

interface NoteObject {
  note_id: string;
  text: string;
  visibility: string;
  owner: string;
}

const VALID_VISIBILITY = new Set(['private', 'party']);

function toNoteObject(row: NoteRow): NoteObject {
  return { note_id: row.note_id, text: row.text, visibility: row.visibility, owner: row.owner };
}

export function createNote(authHeader: string | undefined, campaignId: string, body: JsonValue): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const record = body as Record<string, unknown>;
  const noteId = record?.note_id;
  if (typeof noteId !== 'string' || noteId.length === 0) {
    return { status: 400, body: { error: 'note_id must be a non-empty string' } };
  }
  const text = record?.text;
  if (typeof text !== 'string' || text.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }
  const visibility = record?.visibility;
  if (typeof visibility !== 'string' || !VALID_VISIBILITY.has(visibility)) {
    return { status: 400, body: { error: 'visibility must be "private" or "party"' } };
  }

  const existing = db
    .prepare('SELECT note_id FROM play_campaign_notes WHERE campaign_id = ? AND note_id = ?')
    .get(campaignId, noteId);
  if (existing) {
    return { status: 409, body: { error: 'note_id already exists' } };
  }

  const sequenceRow = db
    .prepare('SELECT COALESCE(MAX(sequence), 0) AS max_sequence FROM play_campaign_notes WHERE campaign_id = ?')
    .get(campaignId) as { max_sequence: number };

  db.prepare(
    'INSERT INTO play_campaign_notes (campaign_id, sequence, note_id, text, visibility, owner) VALUES (?, ?, ?, ?, ?, ?)',
  ).run(campaignId, sequenceRow.max_sequence + 1, noteId, text, visibility, actor.username);

  return {
    status: 201,
    body: { note_id: noteId, text, visibility, owner: actor.username } as unknown as JsonValue,
  };
}

export function listNotes(authHeader: string | undefined, campaignId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const rows = db
    .prepare(
      'SELECT campaign_id, sequence, note_id, text, visibility, owner FROM play_campaign_notes WHERE campaign_id = ? ORDER BY sequence ASC',
    )
    .all(campaignId) as unknown as NoteRow[];

  const isDm = campaign.owner === actor.username;
  const items = rows
    .filter((row) => isDm || row.visibility === 'party' || row.owner === actor.username)
    .map(toNoteObject);

  return { status: 200, body: { notes: items } as unknown as JsonValue };
}

export function getNote(authHeader: string | undefined, campaignId: string, noteId: string): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const row = db
    .prepare('SELECT campaign_id, sequence, note_id, text, visibility, owner FROM play_campaign_notes WHERE campaign_id = ? AND note_id = ?')
    .get(campaignId, noteId) as NoteRow | undefined;
  if (!row) {
    return { status: 404, body: { error: 'note not found' } };
  }

  const isDm = campaign.owner === actor.username;
  if (row.visibility === 'private' && row.owner !== actor.username && !isDm) {
    return { status: 403, body: { error: 'note is private' } };
  }

  return { status: 200, body: toNoteObject(row) as unknown as JsonValue };
}

export function updateNote(
  authHeader: string | undefined,
  campaignId: string,
  noteId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const row = db
    .prepare('SELECT campaign_id, sequence, note_id, text, visibility, owner FROM play_campaign_notes WHERE campaign_id = ? AND note_id = ?')
    .get(campaignId, noteId) as NoteRow | undefined;
  if (!row) {
    return { status: 404, body: { error: 'note not found' } };
  }

  const record = body as Record<string, unknown>;
  const text = record?.text;
  if (typeof text !== 'string' || text.length === 0) {
    return { status: 400, body: { error: 'text must be a non-empty string' } };
  }
  const visibility = record?.visibility;
  if (typeof visibility !== 'string' || !VALID_VISIBILITY.has(visibility)) {
    return { status: 400, body: { error: 'visibility must be "private" or "party"' } };
  }

  if (row.owner !== actor.username) {
    return { status: 403, body: { error: 'only the note owner may update this note' } };
  }

  db.prepare('UPDATE play_campaign_notes SET text = ?, visibility = ? WHERE campaign_id = ? AND note_id = ?').run(
    text,
    visibility,
    campaignId,
    noteId,
  );

  return {
    status: 200,
    body: { note_id: noteId, text, visibility, owner: row.owner } as unknown as JsonValue,
  };
}
